### PR TITLE
Error overhaul

### PR DESCRIPTION
**Pull request by vtjnash**
Fixes for internal (and sometimes external) handling of error codes. This should fix most of the cases where we report the wrong error code, although some cases might still be hanging around (with the most likely culprits now being any sort of logging code).

### CODE REVIEW

File: base/error.jl [hunk 1]
```diff
@@ -145,11 +145,13 @@ end
 
 ## system error handling ##
 """
-    systemerror(sysfunc, iftrue)
+    windowserror(sysfunc[, errno::Cint=Libc.errno()])
```
Comment (JeffBezanson) [inline base/error.jl]:
```suggestion
    systemerror(sysfunc[, errno::Cint=Libc.errno()])
```

File: src/threading.c [hunk 2]
```diff
@@ -113,7 +113,24 @@ BOOLEAN WINAPI DllMain(IN HINSTANCE hDllHandle, IN DWORD nReason,
...
 {
-    return (jl_ptls_t)TlsGetValue(jl_tls_key);
+#if defined(_CPU_X86_64_)
+    DWORD *plast_error = (DWORD*)(__readgsqword(0x30) + 0x68);
+    DWORD last_error = *plast_error;
+#elif defined(_CPU_X86_)
+    DWORD *plast_error = (DWORD*)(__readfsdword(0x18) + 0x34);
+    DWORD last_error = *plast_error;
+#else
+    DWORD last_error = GetLastError();
```
Comment (JeffBezanson) [inline src/threading.c]: Ouch this is an unfortunate design in the windows API...

Comment (vtjnash) [reply src/threading.c]: yeah, and this is not a good place to be having lots of extra function calls, so I'm inlining just the real meat of it (when possible). OTOH, wasn't it great to know that the operation of loading the gc-frame was always successful?

**Comment by vtjnash** [comment]
@Keno can you fix the analyzegc pass for https://build.julialang.org/#/builders/146/builds/3029/steps/3/logs/stdio

**Comment by Keno** [comment]
Just add it to the whitelist: https://github.com/JuliaLang/julia/blob/master/src/clangsa/GCChecker.cpp#L767

**Comment by vtjnash** [comment]
@JeffBezanson could you review 3905102beaba64bb7e19d5308450fc7b93a2e243 (now squashed into here)?

File: src/dlload.c [hunk 3]
```diff
@@ -182,21 +209,30 @@ JL_DLLEXPORT void *jl_load_dynamic_library(const char *modname, unsigned flags,
...
+    for (i = 0; i < n_extensions; i++) {
         const char *ext = extensions[i];
         path[0] = '\0';
         snprintf(path, PATHBUF, "%s%s", modname, ext);
         handle = jl_dlopen(path, flags);
         if (handle)
             goto done;
+#ifdef _OS_WINDOWS_
+        err = GetLastError();
+	break; // LoadLibrary already tested the rest
```
Comment (JeffBezanson) [inline src/dlload.c]: indent

**Review by JeffBezanson — APPROVED** [review]

**Comment by KristofferC** [comment]
Is this really feasible to backport?

**Pull request merged**